File: akka-javasdk/src/main/java/akka/javasdk/consumer/ConsumerProvider.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.javasdk.consumer;

import akka.javasdk.Kalix;
import akka.javasdk.impl.MessageCodec;
import akka.javasdk.impl.consumer.ConsumerRouter;
import com.google.protobuf.Descriptors;

import java.util.Optional;

/**
 * Register an Consumer in {{@link Kalix}} using an <code>ConsumerProvider</code>.
 */
public interface ConsumerProvider<A extends Consumer> {

  ConsumerOptions options();

  Descriptors.ServiceDescriptor serviceDescriptor();

  ConsumerRouter<A> newRouter(ConsumerContext context);

  Descriptors.FileDescriptor[] additionalDescriptors();

  default Optional<MessageCodec> alternativeCodec() {
    return Optional.empty();
  }
}

File: akka-javasdk/src/main/scala/akka/javasdk/impl/consumer/ConsumerProvider.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.javasdk.impl.consumer

import akka.annotation.InternalApi
import akka.javasdk.consumer.Consumer
import akka.javasdk.impl.ComponentDescriptor
import akka.javasdk.impl.ComponentDescriptorFactory
import akka.javasdk.impl.JsonMessageCodec
import com.google.protobuf.Descriptors

/**
 * INTERNAL API
 */
@InternalApi
private[impl] final case class ConsumerProvider[A <: Consumer](
    cls: Class[A],
    messageCodec: JsonMessageCodec,
    factory: () => A) {
  private val componentDescriptor = ComponentDescriptor.descriptorFor(cls, messageCodec)
  val serviceDescriptor: Descriptors.ServiceDescriptor = componentDescriptor.serviceDescriptor

  private def newRouter(): ConsumerRouter[A] = {
    val consumer: A = factory()
    new ReflectiveConsumerRouter[A](
      consumer,
      componentDescriptor.commandHandlers,
      ComponentDescriptorFactory.findIgnore(consumer.getClass))
  }

  def newServiceInstance(): ConsumerService =
    new ConsumerService(
      newRouter _,
      serviceDescriptor,
      Array[Descriptors.FileDescriptor](componentDescriptor.fileDescriptor),
      messageCodec)

}
